api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -140,6 +139,17 @@ message Matcher {
   OnMatch on_no_match = 3;
 }
 
+// Wrapper around an existing extension that provides an associated matcher. This allows
+// decorating an existing extension with a matcher, which can be used to match against
+// relevant protocol data.
+message ExtensionWithMatcher {
```

Comment:
Thinking about how to structure this a bit more I wonder if this should end up being a new filter that is responsible for wrapping another filter with the match tree. My original thinking was to teach the HCM to unwrap this type, but there are quite a few places where this has to be done (static config, dynamic ECDS config, default ECDS config), so it might be nicer to just delegate all this work to a filter factory. This might influence where we want to put this file?

This is based on the difficulties I was having trying to do the handling in the HCM, so I'm just speculating that a matching filter would be the way to go. This filter could then also be responsible for *not* injecting a filter if none is specified, satisfying #14322.

Reply:
Snow and I talked about this offline and I think we agreed this is the approach we are going to go with?

Reply:
Snow, are you planning to make more changes in this PR for this?  The current proto still looks basically the way it was before (which looks fine to me, but you and Matt might have a different idea).

Reply:
Not planning on making any changes to this. The conversation I had with Matt landed on not making this appear as a filter in the API (so *not* what I suggested in the top comment here), so the current proto is what we're thinking